barretenberg/cpp/src/barretenberg/ultra_honk/merge_verifier.cpp

```diff
@@ -13,86 +13,180 @@ namespace bb {
+        verified &= (t_eval_kappa_inv * pow_kappa_minus_one == reversed_t_eval);
     }
 
-    OpeningClaim batched_claim = { { kappa, batched_eval }, batched_commitment };
+    return verified;
+}
+
+bool MergeVerifier::verify_claims(
+    std::vector<typename MergeVerifier::Commitment>& table_commitments,
+    const std::vector<typename ShplonkVerifier::LinearCombinationOfClaims>& opening_claims)
+{
+    // Initialize Shplonk verifier
+    ShplonkVerifier verifier(table_commitments, transcript, NUM_MERGE_CLAIMS);
+    verifier.reduce_verification_vector_claims_no_finalize(opening_claims);
+
+    // Export batched claim
+    auto batch_opening_claim = verifier.export_batch_opening_claim(Commitment::one());
 
-    auto pairing_points = PCS::reduce_verify(batched_claim, transcript);
+    // KZG verifier
+    auto pairing_points = PCS::reduce_verify_batch_opening_claim(batch_opening_claim, transcript);
     VerifierCommitmentKey pcs_vkey{};
     auto verified = pcs_vkey.pairing_check(pairing_points[0], pairing_points[1]);
-    return identity_checked && verified;
+
+    return verified;
+}
+
+/**
+ * @brief Verify proper construction of the aggregate Goblin ECC op queue polynomials T_j, j = 1,2,3,4.
+ * @details Let T_j be the jth column of the aggregate ecc op table after prepending the subtable columns t_j
+ * containing the contribution from a single circuit. T_{j,prev} corresponds to the columns of the aggregate table
+ * at the previous stage. For each column we have the relationship T_j = t_j + right_shift(T_{j,prev}, k), where k
+ * is the length of the subtable columns t_j. This protocol demonstrates that the aggregate ecc op table has been
+ * constructed correctly via:
+ * - the Schwartz-Zippel check:
+ *      \f[ T_j(\kappa) = t_j(\kappa) + \kappa^k * (T_{j,prev}(\kappa)) \f]
+ * - the degree check a la Thakur:
+ *      \f[ x^{l-1} t_j(1/x) = g_j(x) \f]
+ *   where \f$g_j(X) = X^{l-1} t_j(1 / X)\f$.
+ *
+ * @tparam Flavor
+ * @param t_commitments The commitments to t_j read from the transcript by the PG verifier with which the Merge
+ * verifier shares a transcript
+ * @return bool Verification result
+ */
+bool MergeVerifier::verify_proof(const HonkProof& proof, const RefArray<Commitment, NUM_WIRES>& t_commitments)
+{
+    transcript->load_proof(proof);
+
+    // clang-format off
+    /**
+     * Shplonk verification checks the following openings: t_j(1/kappa), p_j(kappa), g_j(kappa)
+     * The polynomials p_j(X) have the form: p_j(X) = t_j(X) - kappa^l T_{prev, j}(X) - T_j(X). Therefore, the verifier
+     * must compute the commitment to p_j starting from the commitments to t_j, T_{prev, j}, T_j. To avoid unnecessary
+     * ECC operations, we set up the Shplonk verifier as follows.
+     *
+     * ShplonkVerifier.commitments = {
+     *      [t_1], .., [t_4], [T_{1,prev}], .., [T_{4, prev}], [T_1], .., [T_4], [g_1], .., [g_4]
+     * }
+     *
+     * And then we open the claims by linearly combining the commitments:
```

Comment:
This matrix is a bit hard to follow when you're having a first look at his, maybe you can present this piece of info in a different way :)